File: leetcode/easy/459_重复的子字符串/459_重复的子字符串.py
class Solution:
    def getNext(self, s: str) -> list:
        n = len(s)
        next_arr = [-1] * n
        j = -1
        for i in range(1, n):
            while j >= 0 and s[i] != s[j + 1]:
                j = next_arr[j]
            if s[i] == s[j + 1]:
                j += 1
            next_arr[i] = j
        return next_arr
    
    def repeatedSubstringPattern(self, s: str) -> bool:
        if len(s) == 0:
            return False
        next_arr = self.getNext(s)
        length = len(s)
        if next_arr[-1] != -1 and length % (length - (next_arr[-1] + 1)) == 0:
            return True
        return False

File: leetcode/easy/459_重复的子字符串/459_重复的子字符串.go
func getNext(s string) []int {
	n := len(s)
	next := make([]int, n)
	for i := 0; i < n; i++ {
		next[i] = -1
	}
	j := -1
	for i := 1; i < n; i++ {
		for j >= 0 && s[i] != s[j+1] {
			j = next[j]
		}
		if s[i] == s[j+1] {
			j++
		}
		next[i] = j
	}
	return next
}

func repeatedSubstringPattern(s string) bool {
	if len(s) == 0 {
		return false
	}
	next := getNext(s)
	length := len(s)
	if next[length-1] != -1 && length%(length-(next[length-1]+1)) == 0 {
		return true
	}
	return false
}